features/frameworks/greentea-client/greentea-client/greentea_serial.h

```diff
@@ -6,12 +6,16 @@
 
 #if DEVICE_SERIAL
 
-#include "RawSerial.h"
+#include "drivers/UnbufferedSerial.h"
 #include "SingletonPtr.h"
 
-class GreenteaSerial : public mbed::RawSerial {
+class GreenteaSerial : public mbed::UnbufferedSerial {
 public:
     GreenteaSerial();
+
+    int printf(const char *format, ...) MBED_PRINTF_METHOD(1, 2);
+
+    int vprintf(const char *format, std::va_list arg);
```

Comment:
I don't think that we need to provide printf/vprintf functions; I searched for usage of these functions in greentea tests and I can't find any. The only place where it is used is `#define utest_printf(...) greentea_serial->printf(__VA_ARGS__)`. If utest needs a locking wrapper then it should implement its own and not depend on Greentea serial class.

Reply:
As it stands, this doesn't make sense because `GreenteaSerial` is not actually using the `UnbufferedSerial` it's based on anyway - it's outputting to `write(STDERR_FILENO)`.

Which is what I've wanted for a while, since https://github.com/ARMmbed/mbed-os/issues/6674

If no-one is relying on having `greentea_serial` available as a `SerialBase` (and I've been told they shouldn't be), then I don't see any point making it a class.

I'd strip back and put the implementation into `greentea_test_env.c`.

Implement `greentea_write_string` and `greentea_putc` and `greentea_getc` directly as `write(STDERR_FILENO)` and `read(STDERR_FILENO)`.

You then do need `utest_printf`, but that can also be a real function in utest.

And nothing is then serial-specific, so can be removed from any `#if DEVICE_SERIAL`.

Reply:
Please review [this](https://github.com/ARMmbed/mbed-os/compare/57da238f7fd2fc409f2306d7d618ddbc20ac0482..1768819088a52372e50606c7513d8d41577f52f2).